### PR TITLE
[Diff-Transformer]add Diff-Transformer model and test

### PR DESCRIPTION
**Pull request by cf-icehzgzh**
#完成 Diff-Transformer 模型在 PaddleFormers 框架的接入，包含模型核心实现、配置定义、分布式配置、分词器适配、AutoConfig 注册以及标准单元测试
#进入测试目录
cd PaddleFormers/tests/transformers/diff_transformer
#运行测试
python -m pytest test_modeling.py -v
#结果截图
图片

### CODE REVIEW

**Comment by a31413510** [comment]
1. auto/modeling.py: 需要添加 ("DiffTransformer", "diff_transformer") 注册
2. transformers/__init__.py: 需要添加 diff_transformer lazy imports
3. diff_transformer/modeling.py:
    - MLP nn.Linear 需要添加 bias_attr=False
    self.mlp = nn.Sequential(
              nn.Linear(config.hidden_size, config.intermediate_size, bias_attr=False),  # ← 加 bias_attr=False
              nn.Silu(),
              nn.Linear(config.intermediate_size, config.hidden_size, bias_attr=False),  # ← 加 bias_attr=False
          )
    - 需要新增 _gen_aoa_config()，所有 Linear 权重加 ^T 转置，参考：
    class DiffTransformerForCausalLM(DiffTransformerPreTrainedModel):
      @classmethod
      def _gen_aoa_config(cls, config):
          """AOA checkpoint conversion config. Linear weights need transpose (PyTorch [out,in] -> Paddle [in,out])."""
          aoa_statements = [
              "model.embed_tokens.weight -> model.embed_tokens.weight",
              "model.norm.weight -> model.norm.weight",
              "lm_head.weight^T -> lm_head.weight",                                   # ← ^T 转置
              "model.layers.$LAYER_ID.norm1.weight -> model.layers.$LAYER_ID.norm1.weight",
              "model.layers.$LAYER_ID.norm2.weight -> model.layers.$LAYER_ID.norm2.weight",
              "model.layers.$LAYER_ID.attn.q_proj.weight^T -> model.layers.$LAYER_ID.attn.q_proj.weight",  # ← ^T
              "model.layers.$LAYER_ID.attn.k_proj.weight^T -> model.layers.$LAYER_ID.attn.k_proj.weight",  # ← ^T
              "model.layers.$LAYER_ID.attn.v_proj.weight^T -> model.layers.$LAYER_ID.attn.v_proj.weight",  # ← ^T
              "model.layers.$LAYER_ID.attn.o_proj.weight^T -> model.layers.$LAYER_ID.attn.o_proj.weight",  # ← ^T
              "model.layers.$LAYER_ID.attn.lambda_q1 -> model.layers.$LAYER_ID.attn.lambda_q1",
              "model.layers.$LAYER_ID.attn.lambda_k1 -> model.layers.$LAYER_ID.attn.lambda_k1",
              "model.layers.$LAYER_ID.attn.lambda_q2 -> model.layers.$LAYER_ID.attn.lambda_q2",
              "model.layers.$LAYER_ID.attn.lambda_k2 -> model.layers.$LAYER_ID.attn.lambda_k2",
              "model.layers.$LAYER_ID.attn.subln.weight -> model.layers.$LAYER_ID.attn.subln.weight",
              "model.layers.$LAYER_ID.mlp.0.weight^T -> model.layers.$LAYER_ID.mlp.0.weight",             # ← ^T
              "model.layers.$LAYER_ID.mlp.2.weight^T -> model.layers.$LAYER_ID.mlp.2.weight",             # ← ^T
          ]
          return {"aoa_statements": aoa_statements}

**Comment by cf-icehzgzh** [comment]
感谢提出问题！我已经按照您的要求完成了所有代码修改，并提交了新的 commit：
1. 在 `auto/modeling.py` 中注册了模型。
2. 在 `__init__.py` 中添加了懒加载。
3. 在 `DiffTransformerBlock` 的 MLP 中添加了 `bias_attr=False`。
4. 在 `DiffTransformerForCausalLM` 中实现了 `_gen_aoa_config` 方法。

**Comment by cf-icehzgzh** [comment]
e4a139c2dffef9ba29c8334341d34682
已修改冲突问题

File: paddleformers/transformers/auto/__init__.py [hunk 1]
```diff
@@ -53,10 +53,22 @@
     "image_processing": ["get_image_processor_config", "AutoImageProcessor", "IMAGE_PROCESSOR_MAPPING"],
     "processing": ["AutoProcessor", "PROCESSOR_MAPPING"],
     "video_processing": ["AutoVideoProcessor", "VIDEO_PROCESSOR_MAPPING"],
+    "diff_transformer": [
```
Comment (risemeup1111) [inline paddleformers/transformers/auto/__init__.py]: P1 **优先级：P1**
**处理要求：请针对该评论修复并提交新的 commit。**

这里把 DiffTransformer 注册到了 `paddleformers.transformers.auto` 的 lazy import 里，`_LazyModule` 会按这个条目去导入 `paddleformers.transformers.auto.diff_transformer`；但当前仓库只有 `paddleformers/transformers/diff_transformer/`，没有 `paddleformers/transformers/auto/diff_transformer.py`。因此访问 `from paddleformers.transformers.auto import DiffTransformerConfig` 这类导入会在运行时失败。同时之前需要补的是顶层 `paddleformers/transformers/__init__.py` 的 lazy imports，当前顶层文件仍没有 `diff_transformer` 条目，常用的 `from paddleformers.transformers import DiffTransformerForCausalLM` 也不会导出。

请移除这里的 `auto.diff_transformer` lazy entry，并在 `paddleformers/transformers/__init__.py` 添加顶层导出，形态可以参考：

```python
"diff_transformer.configuration": ["DiffTransformerConfig"],
"diff_transformer.modeling": [
    "DiffTransformerModel",
    "DiffTransformerForCausalLM",
    "DiffTransformerPreTrainedModel",
],
"diff_transformer.tokenizer": ["DiffTransformerTokenizer"],
"diff_transformer": [],
```

同时在 `TYPE_CHECKING` 分支补 `from .diff_transformer import *`，并建议加一个公开导入测试覆盖 `from paddleformers.transformers import DiffTransformerConfig, DiffTransformerForCausalLM`。

File: paddleformers/transformers/diff_transformer/modeling.py [hunk 2]
```diff
@@ -0,0 +1,164 @@
...
+        bsz, seq_len, _ = hidden_states.shape
+
+        q = self.q_proj(hidden_states).reshape(bsz, seq_len, 2, self.num_heads, self.head_dim)
+        k = self.k_proj(hidden_states).reshape(bsz, seq_len, 2, self.num_heads, self.head_dim)
+        v = self.v_proj(hidden_states).reshape(bsz, seq_len, self.num_heads, self.head_dim)
+
+        q1, q2 = q[:, :, 0], q[:, :, 1]
+        k1, k2 = k[:, :, 0], k[:, :, 1]
+
+        attn1 = F.scaled_dot_product_attention(q1, k1, v, attn_mask=attention_mask)
```
Comment (risemeup1111) [inline paddleformers/transformers/diff_transformer/modeling.py]: P1 **优先级：P1**
**处理要求：请针对该评论修复并提交新的 commit。**

`DiffTransformerForCausalLM` 默认调用时会把 `attention_mask=None` 一路传到这里，但这两次 `scaled_dot_product_attention` 没有设置 `is_causal=True`，也没有由 `DiffTransformerModel.forward` 生成下三角 causal mask。这样训练/推理时每个位置都能 attend 到未来 token，CausalLM 的目标被破坏；当前测试只检查输出非空和 backward，覆盖不到这个泄漏。

请在进入每层 attention 前生成 causal mask，或切到仓库已有的 `ALL_ATTENTION_FUNCTIONS` 并传入 `is_causal`/mask，参考同目录其它 decoder-only 模型的处理。修复形态应类似：

```python
# DiffTransformerModel.forward 中基于 input_ids.shape 生成 [bsz, 1, seq_len, seq_len] 的 causal mask
causal_mask = self._prepare_decoder_attention_mask(
    attention_mask=attention_mask,
    input_shape=input_ids.shape,
    past_key_values_length=0,
    dtype=x.dtype,
)
for layer in self.layers:
    x = layer(x, attention_mask=causal_mask)
```

或者在直接调用 SDPA 时显式使用 `is_causal=attention_mask is None and seq_len > 1`，同时保留外部 padding mask 与 causal mask 的组合逻辑。请补一个测试：修改输入末尾 token 后，较早位置的 logits 不应变化。

File: paddleformers/transformers/diff_transformer/modeling.py [hunk 3]
```diff
@@ -0,0 +1,164 @@
...
+    def forward(self, x):
+        return x * paddle.rsqrt(paddle.mean(x**2, axis=-1, keepdim=True) + self.eps) * self.weight
+
+
+class DiffAttn(nn.Layer):
+    def __init__(self, config, layer_idx):
+        super().__init__()
+        self.hidden_size = config.hidden_size
+        self.num_heads = config.num_attention_heads
+        self.head_dim = self.hidden_size // self.num_heads
```
Comment (risemeup1111) [inline paddleformers/transformers/diff_transformer/modeling.py]: P2 **优先级：P2**
**处理要求：请针对该评论进行回复（同意并已修改请回复 Done，不同意请说明理由）。**

配置类暴露并保存了 `head_dim`，但这里重新用 `hidden_size // num_attention_heads` 计算，导致用户或转换脚本在 config 里传入的 `head_dim` 被静默忽略。类似 `LlamaConfig`/`GptOssConfig` 都会把 `head_dim` 作为可配置字段保存并在模型里使用；如果 DiffTransformer 不支持自定义 head dim，也建议在 config 层去掉该参数或显式校验，避免 checkpoint 配置看起来生效但实际没有生效。

可选修复形态：

```python
self.head_dim = config.head_dim
if self.head_dim * self.num_heads != self.hidden_size:
    raise ValueError("head_dim * num_attention_heads must equal hidden_size for DiffTransformer")
```

如果该模型设计上固定使用 `hidden_size // num_attention_heads`，请移除 `DiffTransformerConfig.head_dim` 或在初始化时明确覆盖并说明原因。

**Review by risemeup1111 — CHANGES_REQUESTED** [review]
已完成首轮评审。当前仍有需要合入前修复的问题，主要集中在 DiffTransformer 的公开导入注册和 CausalLM attention 的因果遮罩；具体修改建议已放在行级评论中。CI 当前显示通过，但这些行为问题不会被现有测试覆盖到。

<div align="right">
   <sup>Powered by Nyanpasu with gpt-5.5 xhigh, please check the suggestions carefully.</sup>
</div>

**Review by risemeup1111 — COMMENTED** [review]
已复查最新提交。公开导入注册和 CausalLM 因果遮罩这两处 P1 问题已按代码改动处理，`head_dim` 也已增加一致性校验和测试。

目前只剩原 `head_dim` 线程需要作者按约定回复（同意并已修改请回复 Done；不同意请说明理由）。收到回复后我再做最终确认。

<div align="right">
   <sup>Powered by Nyanpasu with gpt-5.5 xhigh, please check the suggestions carefully.</sup>
</div>

File: paddleformers/transformers/diff_transformer/modeling.py (hunk 3 above)
Comment (cf-icehzgzh) [reply paddleformers/transformers/diff_transformer/modeling.py]: Done

**Review by risemeup1111 — APPROVED** [review]
已复查最新提交和作者在 `head_dim` 线程中的回复。此前关于公开导入注册、CausalLM 因果遮罩以及 `head_dim` 配置一致性的意见都已处理，未发现需要继续阻塞合入的问题。

本地未运行单测：当前环境缺少 `paddle` 依赖；PR 当前 CI 检查通过。

<div align="right">
   <sup>Powered by Nyanpasu with gpt-5.5 xhigh, please check the suggestions carefully.</sup>
</div>

File: paddleformers/transformers/diff_transformer/modeling.py [hunk 4]
```diff
@@ -0,0 +1,183 @@
...
+        return self.o_proj(attn), None, None
+
+
+class DiffTransformerBlock(nn.Layer):
+    def __init__(self, config, layer_idx):
+        super().__init__()
+        self.norm1 = RMSNorm(config.hidden_size)
+        self.attn = DiffAttn(config, layer_idx)
+        self.norm2 = RMSNorm(config.hidden_size)
+        # 修改：添加 bias_attr=False
```
Comment (a31413510) [inline paddleformers/transformers/diff_transformer/modeling.py]: 需要去掉中文注释

**Comment by a31413510** [comment]
tests/transformers/diff_transformer/__init__.py 缺失

**Comment by a31413510** [comment]
代码风格检测失败，本地可以使用pre-commit自行校验

**Comment by a31413510** [comment]
CI 单测结果 — 1 个 FAIL

  tests/transformers/diff_transformer/test_diff_transformer_modeling.py
    7 passed, 1 FAILED

  FAILED: test_causal_lm_does_not_attend_to_future_tokens

  失败原因：paddle.testing 模块没有 assert_allclose 方法（该方法名在此版本 PaddlePaddle 中不存在），应改为 paddle.testing.assert_close：

  # 第136行，错误写法：
  paddle.testing.assert_allclose(logits[:, :-1], modified_logits[:, :-1])

  # 应改为：
  paddle.testing.assert_close(logits[:, :-1], modified_logits[:, :-1])

**Review by liuhao2638 — APPROVED** [review]
已复查最新提交。该提交仅移除了 `DiffTransformerBlock` 中的中文注释，不影响此前已验证的公开导入注册、CausalLM 因果遮罩和 `head_dim` 配置一致性修复；未发现需要阻塞合入的问题。

本地未运行单测：当前环境缺少 `paddle` 依赖；PR 当前 CI 检查通过。

<div align="right">
   <sup>Powered by Nyanpasu with gpt-5.5 xhigh, please check the suggestions carefully.</sup>
</div>

**Comment by a31413510** [comment]
仍然存在代码风格报错
image
本地commit前，先执行pre-commit解决代码规范问题再提交

File: tests/transformers/diff_transformer/test_diff_transformer_modeling.py [hunk 5]
```diff
@@ -0,0 +1,143 @@
...
+        self.tester.check_causal_lm()
+
+    def test_loss_computation(self):
+        self.tester.check_loss()
+
+    def test_backward_pass(self):
+        self.tester.check_backward()
+
+    def test_public_imports(self):
+        from paddleformers.transformers import (
```
Comment (risemeup1111) [inline tests/transformers/diff_transformer/test_diff_transformer_modeling.py]: P1 **优先级：P1**
**处理要求：请针对该评论修复并提交新的 commit。**

这个格式化提交把 `tests/transformers/diff_transformer/test_diff_transformer_modeling.py` 保存成了 CRLF 行尾：`file tests/transformers/diff_transformer/test_diff_transformer_modeling.py` 显示 `with CRLF line terminators`，并且 `git diff --check 50b1f220b4b8a9342f872bdf1c501f19145df2a6..084d044f5d64329a00a251055859212ee3f26b2d -- tests/transformers/diff_transformer/test_diff_transformer_modeling.py` 仍报告 107-110、139、143 行 trailing whitespace。也就是说当前代码风格检查仍可能失败，之前 reviewer 提到的 pre-commit 问题没有真正消除。

请把该文件统一转回 LF，并重新跑空白/格式检查，例如：

```bash
python - <<'PY'
from pathlib import Path

path = Path("tests/transformers/diff_transformer/test_diff_transformer_modeling.py")
path.write_text(path.read_text().replace("\r\n", "\n"), newline="\n")
PY
git diff --check
pre-commit run --files tests/transformers/diff_transformer/test_diff_transformer_modeling.py
```

**Review by risemeup1111 — CHANGES_REQUESTED** [review]
已复查最新提交。此前关于公开导入注册、CausalLM 因果遮罩和 `head_dim` 配置一致性的意见仍保持已修复；但本次格式化提交仍留下 CRLF/尾随空白，可能继续触发代码风格检查，具体证据和处理建议已放在行级评论中。

本地未运行单测：当前环境缺少 `paddle` 依赖；PR 当前基础检查通过，但本地 `git diff --check` 对该测试文件仍失败。

<div align="right">
   <sup>Powered by Nyanpasu with gpt-5.5 xhigh, please check the suggestions carefully.</sup>
</div>

**Review by risemeup1111 — APPROVED** [review]
已复查最新提交。此前的公开导入注册、CausalLM 因果遮罩和 `head_dim` 配置一致性问题仍保持已修复；本次提交也已将 DiffTransformer 测试文件恢复为 LF 行尾，本地 `git diff --check` 不再报告该文件的尾随空白问题。未发现需要继续阻塞合入的问题。

本地未运行单测：当前环境缺少 `paddle` 依赖；PR 当前基础检查通过。

<div align="right">
   <sup>Powered by Nyanpasu with gpt-5.5 xhigh, please check the suggestions carefully.</sup>
</div>

**Comment by a31413510** [comment]
tests/transformers/diff_transformer/__init__.py 缺失，需要补充

**Review by risemeup1111 — APPROVED** [review]
已复查最新提交。此前关于公开导入注册、CausalLM 因果遮罩、`head_dim` 配置一致性以及测试文件行尾/尾随空白的问题均保持已修复；本次提交补充了 `tests/transformers/diff_transformer/__init__.py`，未发现需要继续阻塞合入的问题。

本地未运行单测：当前环境缺少 `paddle` 依赖；PR 当前基础检查通过。

<div align="right">
   <sup>Powered by Nyanpasu with gpt-5.5 xhigh, please check the suggestions carefully.</sup>
</div>
